doc/design/error_reporting.md

```diff
@@ -0,0 +1,456 @@
+in this case.
+
+Use
+
+```cpp
+IOX_ASSUME(x>=0, "")
+```
+if there is no message.
+
+### Marking Unreachable Code
+
+It is also possible to explcitly state that code is supposed to be unreachable.
+```cpp
+if(condition) {
+    // Reachable code that does something
+    // This also implies that it is assumed that the condition cannot be false.
+} else {
+    IOX_UNREACHABLE();
+    // Code here should be dead, otherwise it is a bug
+    // There should ideally be no dead code, but there are exceptions.
+}
+```
+
+If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 
+`panic` will be invoked and the program aborts. Stating that specific code cannot be reached is
+a specific assumption and any violation is considered a bug (hence checking can be optionally
+disabled).
+
+This has advantages for test coverage as the compiler and other tools that rely on the compiler (say
+for coverage) are aware of the `noreturn` guarantee of `IOX_UNREACHABLE`.
+As a consequnece, branches with `IOX_UNREACHABLE` do not necessarily lead to a return statement.
+
+### Summary
+
+This shows how the API can be used to either signal errors to an underlying backend or safeguard
+against bug conditions such as precodition violations. As the latter should not happen in correct
+code, these can be disabled.
+
+## Examples
+
+The following examples show how non-fatal and fatal errors cane be signaled.
+
+## Recoverable Errors
+
+The default version only supports error codes.
+The following assumes there exists a `RUNTIME_ERROR` category.
+
+```cpp
+enum class Code {
+    SomeError
+};
+
+expected<int, Code> algorithm(int x)
+{
+    if(errorCondition(x))
+    {
+        IOX_REPORT(SomeError, RUNTIME_ERROR);
+        // control flow continues and the error is propagated to the caller
+        return error<Code>(SomeError);
+    }
+    return 42;
+}
+
+expected<int, E> identity() {
+    
+    auto result = algorithm(73);
+
+    if(result.has_error())
+    {
+        // transform the error to E and propagate it
+        return into<error<E>>(result.get_error());
+    }
+    
+    // no error, return identity
+    return *result;
+};
+```
+
+This is similar to exception handling without the convenience of propagation.
+While this shows the use with `expected`, it can be used with any error return type, 
+for example the error code itself.
+
+A generalization allows to report more complex error types directly. This requires a corresponding
+custom implementation.
+
+```cpp
+expected<int, Code> algorithm(int x)
+{
+    if(errorCondition(x))
+    {
+        // create an exception like custom error
+        auto e = error<CustomError>(SomeError, "additional error info");
+        // report e directly
+        IOX_REPORT_FATAL(e, RUNTIME_ERROR);
+        return e;
+    }
+    return 42;
+}
+
+expected<int, E> identity()
+{
+    auto result = algorithm();
```

Comment:
Why? I mean in theory it could be failing due to side effect but I will add an argument. 

Reminds me of a Monty Python sketch actually. I like to have an argument...

Reply:
The naming `identity` was off as well, the example changed but the naming did not. Updated.